### PR TITLE
[-] Core : #PSCFV-11357 password multiple hash and new salt

### PR DESCRIPTION
**Pull request by sfroment**
http://forge.prestashop.com/browse/PSCFV-11357 

normal that travis build don't pass hold php syntax on array declaration

### CODE REVIEW

File: classes/Customer.php [hunk 1]
```diff
@@ -318,24 +314,42 @@ public function getByEmail($email, $passwd = null, $ignore_guest = true)
...
+        '.(isset($passwd) ? 'AND `passwd` = \''.pSQL($hash).'\'' : '').'
+        AND `deleted` = 0
+        '.($ignore_guest ? ' AND `is_guest` = 0' : ''));
 
         if (!$result) {
             return false;
         }
+
+
+        if (!$crypto->isFirstHash($passwd, $hash, _COOKIE_KEY_)) {
```
Comment (xGouley) [inline classes/Customer.php]: $crypto could be undefined: if !isset($passwd)

Comment (sfroment42) [inline classes/Customer.php]: I've had a try catch to avoid the problem !

**Comment by sfroment42** [comment]
need to do namespace after merge @kelu95

File: classes/Customer.php [hunk 2]
```diff
@@ -369,11 +387,11 @@ public static function isBanned($id_customer)
         $cache_id = 'Customer::isBanned_'.(int)$id_customer;
         if (!Cache::isStored($cache_id)) {
             $result = (bool)!Db::getInstance(_PS_USE_SQL_SLAVE_)->getRow('
```
Comment (tiennou) [inline classes/Customer.php]: `(bool)` ?

Comment (tiennou) [inline classes/Customer.php]: Oh sorry, I was looking at some mostly-same-but-unrelated line locally ;-).

**Comment by julienbourdeau** [comment]
@sacha thank you for rebasing this PR after PSR-2 changes :D I literally saved hours and hours !

**Comment by julienbourdeau** [comment]
I just merged it on 1.6

cry

**Comment by julienbourdeau** [comment]
Alright so now it's cherry-picked on `develop`: https://github.com/PrestaShop/PrestaShop/compare/69e3359f2164d251aa77d70fd8e08318c6f5fdb6...032a4a8742f43ad5960974bd16c467c854927638

That was easy so thanks for rebasing :D

**Comment by tiennou** [comment]
Note that I would really like to have a little chat with you guys about this :anguished:. As this PR stands, there are IMHO a few shortcomings in the API — for example, the `firstHash` thing will creep up all over the codebase for no good reason.

FYI, I've just rebased #3888 on top of current `develop`. Feel free to merge only @sfroment42's commits from it if you want (the one you want is 6d8db1f8e48edfbcef68e7f78039442f7533718f in my fork).

**Comment by julienbourdeau** [comment]
Oh bad timing for us. Do you thnk you can rebase your PR without sacha's commits ? Describe what you are doing and why you're doing it and I'll look at it.
Thanks !

**Comment by tiennou** [comment]
I'm waiting for `develop` to settle ;-). You can check Forge's http://forge.prestashop.com/browse/PSCFV-11357 for more insight, I think it just lacks a note about supporting additional hash methods : this is done and working here.

**Comment by tiennou** [comment]
Looks like you cherry-picked changes to 1.6.2.0.sql, not sure if that's what you want.

**Comment by tiennou** [comment]
Rebased #3888.

**Comment by tiennou** [comment]
Rebased #3888. Note that 7c546a5668e430e76285598249105371a429c2c6 is a bugfix to this one, else customer hash get truncated when saved and you break all customer logins.

**Comment by julienbourdeau** [comment]
Thank you for rebasing, I will look at it. The fix seems definitely required.

I fixed SQL file here https://github.com/PrestaShop/PrestaShop/commit/8593720e4a69ea0671e87e9cffc7fb77d84748ac Thanks

**Comment by tiennou** [comment]
Also 5bc358af6ad0c1da571ec84d516620995ce84ab1 fixes an issue I had with running tests because of the duplicated define.
